<a href="https://colab.research.google.com/github/evgeniytr1509/HW_DS_11/blob/main/DS_HW_11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense

num_words = 10000
maxlen = 200

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=num_words)
X_train = pad_sequences(X_train, maxlen=maxlen)
X_test = pad_sequences(X_test, maxlen=maxlen)

17464789/17464789 [==============================] - 0s 0us/step


In [2]:
# Простая рекуррентная нейронная сеть (RNN)
model_rnn = Sequential()
model_rnn.add(Embedding(input_dim=num_words, output_dim=32, input_length=maxlen))
model_rnn.add(SimpleRNN(units=32))
model_rnn.add(Dense(units=1, activation='sigmoid'))
model_rnn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# LSTM
model_lstm = Sequential()
model_lstm.add(Embedding(input_dim=num_words, output_dim=32, input_length=maxlen))
model_lstm.add(LSTM(units=32))
model_lstm.add(Dense(units=1, activation='sigmoid'))
model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Двусторонняя рекуррентная нейронная сеть
model_bidirectional = Sequential()
model_bidirectional.add(Embedding(input_dim=num_words, output_dim=32, input_length=maxlen))
model_bidirectional.add(Bidirectional(LSTM(units=32)))
model_bidirectional.add(Dense(units=1, activation='sigmoid'))
model_bidirectional.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Глубокая сеть с несколькими слоями
model_deep = Sequential()
model_deep.add(Embedding(input_dim=num_words, output_dim=32, input_length=maxlen))
model_deep.add(Bidirectional(LSTM(units=32, return_sequences=True)))
model_deep.add(Bidirectional(LSTM(units=32)))
model_deep.add(Dense(units=1, activation='sigmoid'))
model_deep.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Обучение моделей - 3 эпохи
model_rnn.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)
model_lstm.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)
model_bidirectional.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)
model_deep.fit(X_train, y_train, epochs=3, batch_size=128, validation_split=0.2)

Epoch 1/3
157/157 [==============================] - 14s 78ms/step - loss: 0.5521 - accuracy: 0.7107 - val_loss: 0.4384 - val_accuracy: 0.8072
Epoch 2/3
157/157 [==============================] - 10s 67ms/step - loss: 0.3864 - accuracy: 0.8336 - val_loss: 0.3494 - val_accuracy: 0.8502
Epoch 3/3
157/157 [==============================] - 11s 71ms/step - loss: 0.2611 - accuracy: 0.9000 - val_loss: 0.3816 - val_accuracy: 0.8380
Epoch 1/3
157/157 [==============================] - 29s 166ms/step - loss: 0.5023 - accuracy: 0.7501 - val_loss: 0.3279 - val_accuracy: 0.8672
Epoch 2/3
157/157 [==============================] - 26s 165ms/step - loss: 0.2517 - accuracy: 0.9036 - val_loss: 0.2936 - val_accuracy: 0.8738
Epoch 3/3
157/157 [==============================] - 24s 155ms/step - loss: 0.1811 - accuracy: 0.9338 - val_loss: 0.3171 - val_accuracy: 0.8670
Epoch 1/3
157/157 [==============================] - 48s 280ms/step - loss: 0.5147 - accuracy: 0.7380 - val_loss: 0.3270 - val_accuracy: 0.

**Обучение модели RNN 6 эпох**

In [28]:
model_rnn.fit(X_train, y_train, epochs=6, batch_size=128, validation_split=0.2)

Epoch 1/6
157/157 [==============================] - 12s 78ms/step - loss: 0.0223 - accuracy: 0.9957 - val_loss: 0.5812 - val_accuracy: 0.8172
Epoch 2/6
157/157 [==============================] - 13s 80ms/step - loss: 0.0167 - accuracy: 0.9967 - val_loss: 0.6131 - val_accuracy: 0.8316
Epoch 3/6
157/157 [==============================] - 11s 70ms/step - loss: 0.0116 - accuracy: 0.9976 - val_loss: 0.6400 - val_accuracy: 0.8366
Epoch 4/6
157/157 [==============================] - 12s 74ms/step - loss: 0.0052 - accuracy: 0.9994 - val_loss: 0.7001 - val_accuracy: 0.8302
Epoch 5/6
157/157 [==============================] - 14s 91ms/step - loss: 0.0035 - accuracy: 0.9996 - val_loss: 0.7949 - val_accuracy: 0.7868
Epoch 6/6
157/157 [==============================] - 12s 78ms/step - loss: 0.0525 - accuracy: 0.9830 - val_loss: 0.7349 - val_accuracy: 0.7930


Точность обучения RNN:

3 эпохи - 83,8 %

6 эпох - 79,3 %

максимальная точность 83,66% после 3ей эпохи обучения.

**Обучение модели LSTM 6 эпох**


In [29]:
model_lstm.fit(X_train, y_train, epochs=6, batch_size=128, validation_split=0.2)

Epoch 1/6
157/157 [==============================] - 26s 164ms/step - loss: 0.0623 - accuracy: 0.9823 - val_loss: 0.4856 - val_accuracy: 0.8636
Epoch 2/6
157/157 [==============================] - 24s 151ms/step - loss: 0.0690 - accuracy: 0.9790 - val_loss: 0.4682 - val_accuracy: 0.8462
Epoch 3/6
157/157 [==============================] - 27s 171ms/step - loss: 0.0525 - accuracy: 0.9852 - val_loss: 0.4859 - val_accuracy: 0.8578
Epoch 4/6
157/157 [==============================] - 26s 168ms/step - loss: 0.0582 - accuracy: 0.9826 - val_loss: 0.5049 - val_accuracy: 0.8566
Epoch 5/6
157/157 [==============================] - 26s 163ms/step - loss: 0.0389 - accuracy: 0.9898 - val_loss: 0.6333 - val_accuracy: 0.8486
Epoch 6/6
157/157 [==============================] - 29s 184ms/step - loss: 0.0280 - accuracy: 0.9936 - val_loss: 0.5904 - val_accuracy: 0.8580


Точность обучения LSTM:

3 эпохи - 86.7 %

6 эпох - 85.8 %

максимальная точность 86,36% после 1й эпохи обучения

**Обучение модели  двусторонняя рекуррентная нейронная сеть 6 эпох**

In [30]:
model_bidirectional.fit(X_train, y_train, epochs=6, batch_size=128, validation_split=0.2)

Epoch 1/6
157/157 [==============================] - 45s 284ms/step - loss: 0.0689 - accuracy: 0.9802 - val_loss: 0.4872 - val_accuracy: 0.8642
Epoch 2/6
157/157 [==============================] - 44s 278ms/step - loss: 0.0755 - accuracy: 0.9741 - val_loss: 0.4851 - val_accuracy: 0.8634
Epoch 3/6
157/157 [==============================] - 43s 276ms/step - loss: 0.0522 - accuracy: 0.9850 - val_loss: 0.5642 - val_accuracy: 0.8666
Epoch 4/6
157/157 [==============================] - 44s 278ms/step - loss: 0.0383 - accuracy: 0.9898 - val_loss: 0.5519 - val_accuracy: 0.8576
Epoch 5/6
157/157 [==============================] - 45s 286ms/step - loss: 0.0504 - accuracy: 0.9855 - val_loss: 0.5077 - val_accuracy: 0.8588
Epoch 6/6
157/157 [==============================] - 43s 273ms/step - loss: 0.0478 - accuracy: 0.9857 - val_loss: 0.6517 - val_accuracy: 0.8566


Точность обучения:

3 эпохи - 87.58 %

6 эпох - 85.66 %

максимальная точность 86,66% после 3й эпохи обучения

**Обучение модели DEEP (Глубокая сеть с несколькими слоями) 6 эпох**

In [31]:
model_deep.fit(X_train, y_train, epochs=6, batch_size=128, validation_split=0.2)

Epoch 1/6
157/157 [==============================] - 93s 594ms/step - loss: 0.0608 - accuracy: 0.9818 - val_loss: 0.7014 - val_accuracy: 0.8200
Epoch 2/6
157/157 [==============================] - 93s 596ms/step - loss: 0.0495 - accuracy: 0.9857 - val_loss: 0.5342 - val_accuracy: 0.8598
Epoch 3/6
157/157 [==============================] - 99s 631ms/step - loss: 0.0400 - accuracy: 0.9894 - val_loss: 0.5336 - val_accuracy: 0.8572
Epoch 4/6
157/157 [==============================] - 92s 590ms/step - loss: 0.0290 - accuracy: 0.9930 - val_loss: 0.6654 - val_accuracy: 0.8372
Epoch 5/6
157/157 [==============================] - 90s 574ms/step - loss: 0.0349 - accuracy: 0.9904 - val_loss: 0.6091 - val_accuracy: 0.8516
Epoch 6/6
157/157 [==============================] - 92s 584ms/step - loss: 0.0269 - accuracy: 0.9937 - val_loss: 0.6656 - val_accuracy: 0.8462


Точность обучения:

3 эпохи - 86.68 %

6 эпох - 84,62 %

максимальная точность 85,98% после 2й эпохи обучения

**Параметры сетей**

Простая рекуррентная нейронная сеть (RNN)

In [3]:
model_rnn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 200, 32)           320000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 32)                2080      
                                                                 
 dense (Dense)               (None, 1)                 33        
                                                                 
Total params: 322113 (1.23 MB)
Trainable params: 322113 (1.23 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


LSTM,  Long Short-Term Memory

In [4]:
model_lstm.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_1 (Embedding)     (None, 200, 32)           320000    
                                                                 
 lstm (LSTM)                 (None, 32)                8320      
                                                                 
 dense_1 (Dense)             (None, 1)                 33        
                                                                 
Total params: 328353 (1.25 MB)
Trainable params: 328353 (1.25 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


Mодель с двусторонней рекуррентной нейронной сетью (Bidirectional LSTM)

In [5]:
model_bidirectional.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, 200, 32)           320000    
                                                                 
 bidirectional (Bidirection  (None, 64)                16640     
 al)                                                             
                                                                 
 dense_2 (Dense)             (None, 1)                 65        
                                                                 
Total params: 336705 (1.28 MB)
Trainable params: 336705 (1.28 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


Глубокая сеть с несколькими слоями

In [6]:
model_deep.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_3 (Embedding)     (None, 200, 32)           320000    
                                                                 
 bidirectional_1 (Bidirecti  (None, 200, 64)           16640     
 onal)                                                           
                                                                 
 bidirectional_2 (Bidirecti  (None, 64)                24832     
 onal)                                                           
                                                                 
 dense_3 (Dense)             (None, 1)                 65        
                                                                 
Total params: 361537 (1.38 MB)
Trainable params: 361537 (1.38 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


**Сравнительный анализ различных типов рекуррентных нейронных сетей**

  На протяжении обучения все модели показывают улучшение точности на валидационных данных с каждой эпохой.

  В начале обучения все модели достигают относительно высокой точности на валидационных данных.

  LSTM и Bidirectional LSTM в целом демонстрируют хорошую точность на протяжении всего обучения.

  В глубокой сети с несколькими слоями после 2-3 эпох точность остается на уровне примерно 0.85, что может свидетельствовать о наличии переобучения.

Исходя из предоставленных результатов на валидационных данных (val_accuracy), наилучшие показатели точности у Bidirectional LSTM